extends DirectionalLight3D

const DAY_LENGTH := 600.0 # Ú©Ù„ Ù…Ø¯Øª Ø±ÙˆØ²ØŒ Ø«Ø§Ù†ÛŒÙ‡
var rotation_speed := 360.0 / DAY_LENGTH

@onready var env := get_world_3d().environment

func _process(delta: float) -> void:
	# Ú†Ø±Ø®Ø´ Ø®ÙˆØ±Ø´ÛŒØ¯
	rotation_degrees.x -= rotation_speed * delta
	if rotation_degrees.x <= -360.0:
		rotation_degrees.x += 360.0

	# Ù…Ø­Ø§Ø³Ø¨Ù‡ Ø²Ø§ÙˆÛŒÙ‡ ÙØ¹Ù„ÛŒ (Ø¯Ø± Ø¨Ø§Ø²Ù‡ -360 ØªØ§ 0)
	var angle = fmod(rotation_degrees.x, 360.0)
	if angle > 0:
		angle -= 360.0
	var x_rot_deg := rotation_degrees.x
	
	# Ø¯Ø± Ù…Ø­Ø¯ÙˆØ¯Ù‡ -180 ØªØ§ 0 -> Ø§Ù†Ø±Ú˜ÛŒ = 0
	if x_rot_deg >= -180.0 and x_rot_deg <= 0.0:
		light_energy = 1
	else:
		light_energy = 0
	# Ù…Ø­Ø§Ø³Ø¨Ù‡â€ŒÛŒ Ø§Ù†Ø±Ú˜ÛŒ Ø¢Ø³Ù…Ø§Ù† Ø¨Ø± Ø§Ø³Ø§Ø³ Ø²Ø§ÙˆÛŒÙ‡
	var energy := _calculate_sky_energy(angle)
	env.background_energy_multiplier = energy
		# ðŸ”† Ù…Ø­Ø§Ø³Ø¨Ù‡â€ŒÛŒ Ø§Ù†Ø±Ú˜ÛŒ Ø®ÙˆØ±Ø´ÛŒØ¯ Ø¨Ø± Ø§Ø³Ø§Ø³ Ø²Ø§ÙˆÛŒÙ‡ (Ø§ÙØ²ÙˆØ¯Ù‡â€ŒØ´Ø¯Ù‡)
func _calculate_sky_energy(angle: float) -> float:
	# Ø·Ù„ÙˆØ¹: -15 ØªØ§ 0 â†’ 0 ØªØ§ 1
	if angle >= -15 and angle <= 0:
		return lerp(1, 0, (angle + 15.0) /15.0)

	
	# Ø±ÙˆØ²: 0 ØªØ§ -165 â†’ 1
	elif angle < 0.0 and angle > -165.0:
		return 1.0
	# ØºØ±ÙˆØ¨: -165 ØªØ§ -180 â†’ 1 ØªØ§ 0
	elif angle <= -165.0 and angle >= -180.0:
		return lerp(1.0, 0.0, (-165.0 - angle) / 15.0)

	# Ø´Ø¨: -180 ØªØ§ -360 â†’ 0
	else:
		return 0.0
